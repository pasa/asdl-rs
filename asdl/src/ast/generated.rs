//This file is auotgenerated by using `cargo gen-syntax`

#![cfg_attr(rustfmt, rustfmt_skip)]

#[derive(PartialEq, Eq, Hash, Debug)]
pub enum Type<'a> {
    SumType(SumType<'a>),
    ProdType(ProdType<'a>),
}
impl<'a> From<SumType<'a>> for Type<'a> {
    fn from(n: SumType) -> Type {
        Type::SumType(n)
    }
}
impl<'a> From<ProdType<'a>> for Type<'a> {
    fn from(n: ProdType) -> Type {
        Type::ProdType(n)
    }
}


#[derive(PartialEq, Eq, Hash, Debug)]
pub struct SumType<'a> {
    pub type_id: TypeId<'a>,
    pub constructors: Vec<Constr<'a>>,
    pub attrs: Option<Attrs<'a>>,
    pub comments: Vec<&'a str>,
}
impl<'a> SumType<'a> {

    pub(crate) fn new(type_id: TypeId<'a>, constructors: Vec<Constr<'a>>, attrs: Option<Attrs<'a>>, comments: Vec<&'a str>) -> Self {
        SumType{ type_id, constructors, attrs, comments }
    }
}

#[derive(PartialEq, Eq, Hash, Debug)]
pub struct ProdType<'a> {
    pub type_id: TypeId<'a>,
    pub fields: Vec<Field<'a>>,
    pub comments: Vec<&'a str>,
}
impl<'a> ProdType<'a> {

    pub(crate) fn new(type_id: TypeId<'a>, fields: Vec<Field<'a>>, comments: Vec<&'a str>) -> Self {
        ProdType{ type_id, fields, comments }
    }
}
#[derive(PartialEq, Eq, Hash, Debug)]
pub enum Field<'a> {
    Required(Required<'a>),
    Optional(Optional<'a>),
    Repeated(Repeated<'a>),
}
impl<'a> From<Required<'a>> for Field<'a> {
    fn from(n: Required) -> Field {
        Field::Required(n)
    }
}
impl<'a> From<Optional<'a>> for Field<'a> {
    fn from(n: Optional) -> Field {
        Field::Optional(n)
    }
}
impl<'a> From<Repeated<'a>> for Field<'a> {
    fn from(n: Repeated) -> Field {
        Field::Repeated(n)
    }
}


#[derive(PartialEq, Eq, Hash, Debug)]
pub struct Required<'a> {
    pub type_id: TypeId<'a>,
    pub id: Option<Id<'a>>,
}
impl<'a> Required<'a> {

    pub(crate) fn new(type_id: TypeId<'a>, id: Option<Id<'a>>) -> Self {
        Required{ type_id, id }
    }
}

#[derive(PartialEq, Eq, Hash, Debug)]
pub struct Optional<'a> {
    pub type_id: TypeId<'a>,
    pub id: Option<Id<'a>>,
}
impl<'a> Optional<'a> {

    pub(crate) fn new(type_id: TypeId<'a>, id: Option<Id<'a>>) -> Self {
        Optional{ type_id, id }
    }
}

#[derive(PartialEq, Eq, Hash, Debug)]
pub struct Repeated<'a> {
    pub type_id: TypeId<'a>,
    pub id: Option<Id<'a>>,
}
impl<'a> Repeated<'a> {

    pub(crate) fn new(type_id: TypeId<'a>, id: Option<Id<'a>>) -> Self {
        Repeated{ type_id, id }
    }
}


#[derive(PartialEq, Eq, Hash, Debug)]
pub struct Root<'a> {
    pub types: Vec<Type<'a>>,
    pub comments: Vec<&'a str>,
}
impl<'a> Root<'a> {

    pub(crate) fn new(types: Vec<Type<'a>>, comments: Vec<&'a str>) -> Self {
        Root{ types, comments }
    }
}

#[derive(PartialEq, Eq, Hash, Debug)]
pub struct Constr<'a> {
    pub id: ConstrId<'a>,
    pub fields: Vec<Field<'a>>,
    pub comments: Vec<&'a str>,
}
impl<'a> Constr<'a> {

    pub(crate) fn new(id: ConstrId<'a>, fields: Vec<Field<'a>>, comments: Vec<&'a str>) -> Self {
        Constr{ id, fields, comments }
    }
}

#[derive(PartialEq, Eq, Hash, Debug)]
pub struct Attrs<'a> {
    pub fields: Vec<Field<'a>>,
}
impl<'a> Attrs<'a> {

    pub(crate) fn new(fields: Vec<Field<'a>>) -> Self {
        Attrs{ fields }
    }
}
#[derive(PartialEq, Eq, Hash, Debug)]
pub struct TypeId<'a>(pub(crate) &'a str);

impl<'a> ToString for TypeId<'a> {

    #[allow(dead_code)]
    fn to_string(&self) -> String {
        self.0.to_string()
    }
}
#[derive(PartialEq, Eq, Hash, Debug)]
pub struct ConstrId<'a>(pub(crate) &'a str);

impl<'a> ToString for ConstrId<'a> {

    #[allow(dead_code)]
    fn to_string(&self) -> String {
        self.0.to_string()
    }
}
#[derive(PartialEq, Eq, Hash, Debug)]
pub struct Id<'a>(pub(crate) &'a str);

impl<'a> ToString for Id<'a> {

    #[allow(dead_code)]
    fn to_string(&self) -> String {
        self.0.to_string()
    }
}