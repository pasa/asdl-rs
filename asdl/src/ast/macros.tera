{%- macro type(ty) -%}
{% set type_name = ty.id | camel %}
#[derive(PartialEq, Eq, Hash, Debug)]
pub struct {{type_name}}<'a> {
    {%- for f in ty.fields %}
    {{ self::field(f = f) }},
    {%- endfor %}
}
impl<'a> {{type_name}}<'a> {

    pub(crate) fn new({{ self::args(fs = ty.fields) }}) -> Self {
        {{type_name}}{ {{ self::names(fs = ty.fields) }} }
    }
}
{%- endmacro type -%}

{%- macro names(fs) -%}
{% for f in fs -%}
{{ f.id | snake  }}
{%- if not loop.last -%}, {% endif -%}
{%- endfor -%}
{%- endmacro args -%}

{%- macro args(fs) -%}
{% for f in fs -%}
{{ self::arg(f = f) }}
{%- if not loop.last -%}, {% endif -%}
{%- endfor -%}
{%- endmacro args -%}

{%- macro arg(f) -%}
{%- set field_name = f.id | snake -%}
    {{field_name}}: {{ self::field_type(f = f) }}
{%- endmacro field -%}

{%- macro field(f) -%}
{%- set field_name = f.id | snake -%}
    pub {{field_name}}: {{ self::field_type(f = f) }}
{%- endmacro field -%}

{%- macro field_type(f) -%}
    {%- if f.is_optional -%}
    Option<{{ self::rust_type(name = f.type_id) }}>
    {%- elif f.is_repeated -%}
    Vec<{{ self::rust_type(name = f.type_id) }}>
    {%- else -%}
    {{ self::rust_type(name = f.type_id) }}
    {%- endif -%}
{%- endmacro field -%}

{%- macro rust_type(name) -%}
    {%- if name == 'str'  -%}
        &'a str
    {%- else -%}
        {{ name | camel }}<'a>
    {%- endif -%}
{%- endmacro rust_type -%}

{%- macro id_type(ty) -%}
{%- set type_name = ty.id | camel -%}
{%- set filed = ty.fields.0 -%}
{%- set field_type = self::field_type(f = filed) -%}
{%- set method_name = filed.id | snake -%}
#[derive(PartialEq, Eq, Hash, Debug)]
pub struct {{type_name}}<'a>(pub(crate) {{ field_type }});

impl<'a> ToString for {{type_name}}<'a> {

    #[allow(dead_code)]
    fn to_string(&self) -> String {
        self.0.to_string()
    }
}
{%- endmacro type -%}

