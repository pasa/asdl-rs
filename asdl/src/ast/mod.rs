mod generated;

pub(crate) use generated::*;
