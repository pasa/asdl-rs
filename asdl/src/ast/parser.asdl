root = (type* types, str* comments)
type = SumType(typeId, constr* constructors, attrs?, str* comments)
    | ProdType(typeId, field* fields, str* comments)
constr = (constrId id, field* fields, str* comments)
field = Required(typeId, id?)
    | Optional(typeId, id?)
    | Repeated(typeId, id?)
attrs = (field* fields)
typeId = (str id)
constrId = (str id)
id = (str id)
