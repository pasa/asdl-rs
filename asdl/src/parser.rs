use nom::IResult;
use std::str;

use super::ast::*;

use std::ops::{RangeFrom, RangeTo};

use nom::{Slice, InputIter, Offset, AsChar, InputTakeAtPosition};
use nom::error::ParseError;
use nom::character::is_alphanumeric;
use nom::character::complete::{char, one_of};
use nom::sequence::{pair, tuple};
use nom::bytes::complete::{take_while_m_n, take_while, is_a};
use nom::branch::alt;
use nom::multi::{separated_list, many0, separated_nonempty_list};
use nom::combinator::{map, opt, recognize};
use nom::character::complete::{multispace0, multispace1, line_ending, not_line_ending, space0};

pub(crate) fn parse(i: &str) -> IResult<&str, Root> {
    map(
        tuple((multispace0, opt(pair(comments, multispace1)), many0(ty))),
        |(_, comments, types)| Root::new(types, comments.map(|(c, _)| c).unwrap_or_default()),
    )(i)
}

fn comment_line(i: &str) -> IResult<&str, &str> {
    let (i, (_, _, _, comment)) = tuple((space0, is_a("//"), space0, not_line_ending))(i)?;
    Ok((i, comment))
}

fn comments(i: &str) -> IResult<&str, Vec<&str>> {
    separated_list(line_ending, comment_line)(i)
}

fn ty(i: &str) -> IResult<&str, Type> {
    alt((map(prod_type, |t| t.into()), map(sum_type, |t| t.into())))(i)
}

fn prod_type(i: &str) -> IResult<&str, ProdType> {
    map(
        tuple((comments, multispace0, type_id, char_ms0('='), fields)),
        |(comments, _, type_id, _, fields)| ProdType::new(type_id, fields, comments),
    )(i)
}

fn sum_type(i: &str) -> IResult<&str, SumType> {
    map(
        tuple((
            multispace0,
            comments,
            opt(line_ending),
            space0,
            type_id,
            char_ms0('='),
            constructors,
            opt(attrs),
        )),
        |(_, comments, _, _, type_id, _, constructors, attrs)| {
            SumType::new(type_id, constructors, attrs, comments)
        },
    )(i)
}

fn attrs(i: &str) -> IResult<&str, Attrs> {
    map(pair(is_a("attributes"), fields), |(_, flds)| Attrs::new(flds))(i)
}

fn constructors(i: &str) -> IResult<&str, Vec<Constr>> {
    separated_nonempty_list(char_ms0('|'), constructor)(i)
}

fn constructor(i: &str) -> IResult<&str, Constr> {
    map(
        tuple((multispace0, comments, multispace0, con_id, multispace0, opt(fields))),
        |(_, comments, _, con_id, _, fields)| {
            Constr::new(con_id, fields.unwrap_or_else(|| vec![]), comments)
        },
    )(i)
}

fn fields(i: &str) -> IResult<&str, Vec<Field>> {
    let fields = separated_list(char_ms0(','), field);
    let (i, (_, fields, _)) = tuple((char_ms0('('), fields, char_ms0(')')))(i)?;
    Ok((i, fields))
}

fn field(i: &str) -> IResult<&str, Field> {
    let (i, (type_id, rep, name)) = tuple((type_id, field_arity, opt(pair(multispace1, id))))(i)?;
    let name = name.map(|n| n.1);
    if let Some(rep) = rep {
        match rep {
            '*' => Ok((i, Repeated::new(type_id, name).into())),
            '?' => Ok((i, Optional::new(type_id, name).into())),
            _ => unreachable!(),
        }
    } else {
        Ok((i, Required::new(type_id, name).into()))
    }
}

fn type_id(i: &str) -> IResult<&str, TypeId> {
    map(
        recognize(pair(
            take_while_m_n(1, 1, is_lowercase),
            take_while(is_alphanumeric_or_underscore),
        )),
        TypeId,
    )(i)
}

fn con_id(i: &str) -> IResult<&str, ConstrId> {
    map(
        recognize(pair(
            take_while_m_n(1, 1, is_uppercase),
            take_while(is_alphanumeric_or_underscore),
        )),
        ConstrId,
    )(i)
}

fn id(i: &str) -> IResult<&str, Id> {
    map(
        recognize(pair(take_while_m_n(1, 1, is_alpha), take_while(is_alphanumeric_or_underscore))),
        Id,
    )(i)
}

fn field_arity(i: &str) -> IResult<&str, Option<char>> {
    opt(one_of("*?"))(i)
}

fn is_alpha(a: char) -> bool {
    is_alphanumeric(a as u8)
}

fn is_uppercase(a: char) -> bool {
    a.is_uppercase()
}

fn is_lowercase(a: char) -> bool {
    a.is_lowercase()
}

pub fn char_ms0<I, E: ParseError<I>>(c: char) -> impl Fn(I) -> IResult<I, I, E>
where
    I: Slice<RangeFrom<usize>>
        + Slice<RangeTo<usize>>
        + Offset
        + InputIter
        + InputTakeAtPosition
        + Clone,
    <I as InputIter>::Item: AsChar + Clone,
    <I as InputTakeAtPosition>::Item: AsChar + Clone,
{
    recognize(tuple((multispace0, char(c), multispace0)))
}

fn is_alphanumeric_or_underscore(a: char) -> bool {
    is_alphanumeric(a as u8) || a == '_'
}

#[cfg(test)]
mod tests {

    use super::*;
    use nom;
    use nom::error::ErrorKind;
    use nom::Err;
    use insta::assert_debug_snapshot_matches;

    #[test]
    fn parse_type_id() {
        assert_eq!(type_id("aBcd1_Efg"), Ok(("", TypeId("aBcd1_Efg"))));
        assert_eq!(type_id("aBcd1,Efg"), Ok((",Efg", TypeId("aBcd1"))));
        assert_eq!(type_id("ABcd1Efg"), Err(Err::Error(("ABcd1Efg", ErrorKind::TakeWhileMN))));
    }

    #[test]
    fn parse_con_id() {
        assert_eq!(con_id("ABcd1_Efg"), Ok(("", ConstrId("ABcd1_Efg"))));
        assert_eq!(con_id("ABcd1,Efg"), Ok((",Efg", ConstrId("ABcd1"))));
        assert_eq!(con_id("aBcd1Efg"), Err(Err::Error(("aBcd1Efg", ErrorKind::TakeWhileMN))));
    }

    #[test]
    fn parse_id() {
        assert_eq!(id("ABcd1_Efg"), Ok(("", Id("ABcd1_Efg"))));
        assert_eq!(id("ABcd1,Efg"), Ok((",Efg", Id("ABcd1"))));
        assert_eq!(id("aBcd1_Efg"), Ok(("", Id("aBcd1_Efg"))));
        assert_eq!(id("aBcd1,Efg"), Ok((",Efg", Id("aBcd1"))));
        assert_eq!(id("_aBcd1Efg"), Err(Err::Error(("_aBcd1Efg", ErrorKind::TakeWhileMN))));
    }

    #[test]
    fn parse_field_arity() {
        assert_eq!(field_arity("? abcd"), Ok((" abcd", Some('?'))));
        assert_eq!(field_arity("* abcd"), Ok((" abcd", Some('*'))));
        assert_eq!(field_arity(" abcd"), Ok((" abcd", None)));
    }

    #[test]
    fn parse_field() {
        assert_eq!(field("type,"), Ok((",", Required::new(TypeId("type"), None).into())));
        assert_eq!(field("type?,"), Ok((",", Optional::new(TypeId("type"), None).into())));
        assert_eq!(field("type*,"), Ok((",", Repeated::new(TypeId("type"), None).into())));

        assert_eq!(
            field("type  name,"),
            Ok((",", Required::new(TypeId("type"), Some(Id("name"))).into()))
        );
        assert_eq!(
            field("type?  name,"),
            Ok((",", Optional::new(TypeId("type"), Some(Id("name"))).into()))
        );
        assert_eq!(
            field("type*  name,"),
            Ok((",", Repeated::new(TypeId("type"), Some(Id("name"))).into()))
        );
    }

    #[test]
    fn parse_fields() {
        assert_eq!(
            fields(" ( type1, type2? name  ) "),
            Ok((
                "",
                vec![
                    Required::new(TypeId("type1"), None).into(),
                    Optional::new(TypeId("type2"), Some(Id("name"))).into()
                ]
            ))
        );
    }

    #[test]
    fn parse_constructor() {
        assert_eq!(
            constructor("ConstrId( type1, type2? name  ) "),
            Ok((
                "",
                Constr::new(
                    ConstrId("ConstrId"),
                    vec![
                        Required::new(TypeId("type1"), None).into(),
                        Optional::new(TypeId("type2"), Some(Id("name"))).into()
                    ],
                    vec![]
                )
            ))
        );

        assert_eq!(
            constructor("ConstrId"),
            Ok(("", Constr::new(ConstrId("ConstrId"), vec![], vec![])))
        );
    }

    #[test]
    fn parse_constructors() {
        assert_eq!(
            constructors("ConstrId1( type1, type2? name  ) | ConstrId2"),
            Ok((
                "",
                vec![
                    Constr::new(
                        ConstrId("ConstrId1"),
                        vec![
                            Required::new(TypeId("type1"), None).into(),
                            Optional::new(TypeId("type2"), Some(Id("name"))).into()
                        ],
                        vec![]
                    ),
                    Constr::new(ConstrId("ConstrId2"), vec![], vec![])
                ]
            ))
        );
    }

    #[test]
    fn parse_comment_line() {
        let asdl = r#"  //comment line1"#;
        assert_eq!(comment_line(asdl), Ok(("", "comment line1")));
    }

    #[test]
    fn parse_comments() {
        let asdl = r#"  // comment line1
                        // comment line2"#;
        assert_eq!(comments(asdl), Ok(("", vec!["comment line1", "comment line2"])));
    }

    #[test]
    fn parse_constructors_with_comments() {
        let asdl = r#"  // ConstrId1 comment
                        ConstrId1( type1, type2? name  ) |
                        // ConstrId2 comment line1
                        // ConstrId2 comment line2
                        ConstrId2"#;
        assert_debug_snapshot_matches!(constructors(asdl));
    }

    #[test]
    fn parse_sum_type() {
        let asdl = r#"
                    // SumType comment line 1
                    // SumType comment line 2
                    sumType = 
                        // ConstrId1 comment
                        ConstrId1( type1, type2? name  ) |
                        // ConstrId2 comment line1
                        // ConstrId2 comment line2
                        ConstrId2"#;
        assert_debug_snapshot_matches!(sum_type(asdl));
    }

    #[test]
    fn parse_prod_type() {
        let asdl = r#"  // prodType comment line 1
                        // prodType comment line 2
                        prodType = ( type1, type2? name  )"#;
        assert_debug_snapshot_matches!(prod_type(asdl));
    }

    #[test]
    fn parse_empty_asdl() {
        let asdl = "";
        assert_debug_snapshot_matches!(parse(asdl));
    }
}
