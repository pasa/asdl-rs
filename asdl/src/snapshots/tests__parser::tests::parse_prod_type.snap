---
created: "2019-09-04T14:39:19.728980290Z"
creator: insta@0.8.1
source: asdl/src/parser.rs
expression: prod_type(asdl)
---
Ok(
    (
        "",
        ProdType {
            type_id: TypeId(
                "prodType",
            ),
            fields: [
                Required(
                    Required {
                        type_id: TypeId(
                            "type1",
                        ),
                        id: None,
                    },
                ),
                Optional(
                    Optional {
                        type_id: TypeId(
                            "type2",
                        ),
                        id: Some(
                            Id(
                                "name",
                            ),
                        ),
                    },
                ),
            ],
            comments: [
                "prodType comment line 1",
                "prodType comment line 2",
            ],
        },
    ),
)
